### ISSUE TITLE
Invalid files followed by "input alignment file does not exist"

### ISSUE THREAD

**Issue by ak-watson** [BenoitMorel/GeneRax#55]
I'm having a strange problem that may well just be a simple mistake on my part, but struggling to find the root cause. 

I am trying to run speciesrax as follows:

~/path/to/generax -s MiniNJ -f families.txt -r UndatedDTL -p speciesrax --prune-species-tree --per-family-rates --si-strategy HYBRID --si-estimate-bl --si-quartet-support

families.txt was generated using build_families_file.py as described in the Wiki.

families.txt looks like this:
[FAMILIES]
- FAMILY1
alignment = /path/to/FAMILY1.phy
mapping =  /path/to/FAMILY1.mapping.tsv
subst_model = LG+G

In this case, my mapping file is in a tab-separated gene to species format. 

When I try to run generax, the first alignment file is detected as invalid, and all following alignment files are marked as not existing (see below for an example). 
However, I have manually checked that path to the 'missing files' is correct. 
I read in a previous issue that generax reads phylip files in the same way as raxml, so I checked the alignments in raxml and didn't have any issues or errors thrown up.  

Briefly, my alignments were generated using muscle, trimmed using BGME, and then checked in python to manually filter out some sequences / alignments from the dataset. 
 
The final phylip files were written by SeqIO using the "phylip-relaxed" format.

As a test, after having issues at first, I used readal to convert my SeqIO output to phylip/phylip4 format, but no luck here either. 

Sequence names don't include any banned characters but do include several underscores.

The number of sites and species at the top of the phylip file seems to be correct. 

I have ran this build of generax (2.0.1) in the exact same way for a different dataset without any problems, so it has to be an issue with my dataset. I'm just not sure what!

Thanks,
Andrew 

Gene tree correction information:
- Gene tree strategy: SPR
- Max gene SPR radius: 5

[00:00:00] Filtering invalid families...

Error in family FAMILY1: Cannot read alignment file (file exists but is invalid)
Error in family FAMILY2: Input alignment file does not exist
Error in family FAMILY3: Input alignment file does not exist
Error in family FAMILY4: Input alignment file does not exist

**Comment by BenoitMorel** [comment]
Hi Andrew,

Could you please provide your family file as well as a few alignment/mapping files, such that I can give it a try?

Best,
Benoit

**Comment by ak-watson** [comment]
Hi Benoit,

I continued checking things last night. Sorry if this is hard to follow.

When trying to put together a small subset of the data "phylip-relaxed" formatted data for you (with 4 alignments) things worked fine.

Generax ran right through to a different error at the end:

00:04:10] [Species BL estimation] Infering reconciliation scenarios
[00:04:10] [Species BL estimation] Infering branch lengths from gene trees
[00:04:10] [Species BL estimation] Done
terminate called after throwing an instance of 'LibpllException'
  what():  Could not load open newick file __random__

At first, I thought that might mean that some of the phylip outputs are ok, but not all of them.

However, I couldn't reproduce the original error with a small subset of the data, even when including the files that were marked as "invalid" in the full dataset.

The error that I'm seeing still occurs with the full dataset in "phylip-relaxed" format.

But exporting the data to fasta instead works fine.

So my best guess is that there is a specific problem with the phylip-relaxed format that is causing issues somewhere, but I'm still not sure exactly where.

All sequence names are in the format COG0001_1 (where the number after the underscore is a unique id).

One thing, at the end of the "Input alignment file does not exist" messages I see the following:

WARNING!!! Found 2268 invalid families (they will be discarded from the analysis)
[00:00:00] Starting species tree initialization...
generax: /home/andrew/Programs/GeneRax/src/core/NJ/NeighborJoining.cpp:58: Position findMinPosition(const DistanceMatrix&): Assertion `minPosition != invalidPosition' failed.
 
So it is only throwing away 2268 families out of 3285.

I have attached the families file along with a few alignments and map files as an example, but like I said, I have had trouble recreating the error with subsets of the dataset.

Thanks,
Andrew

[subset.zip](https://github.com/BenoitMorel/GeneRax/files/6670312/subset.zip)
[families.txt](https://github.com/BenoitMorel/GeneRax/files/6670297/families.txt)

**Comment by BenoitMorel** [comment]
Dear Andrew,

This seems to be the same issue as here:
https://groups.google.com/g/generaxusers/c/-xWj2gmngqI
If you compiled GeneRax from github, you just need to git pull and
recompile to have the fix.
If you installed GeneRax with bioconda, you need to be patient because I
haven't updated the package yet. But I will start the update process this
afternoon, so it should be available soon (hopefully on Monday if the
bioconda team is reactive!).

Alternatively, you can generate the gene trees with ParGenes (see the
discussion in the link above), and then specify them as starting gene trees
in the family file, which will not trigger the bug.

Let me know if you installed GeneRax with bioconda, such that I can notify
you when the package is updated.

Best,
Benoit

**Comment by ak-watson** [comment]
Dear Benoit,

Thanks for taking the time to look at this,

I compiled from github because I wanted to test out some of the new speciesrax functions.

I pulled and recompiled but unfortunately have the same problem as before (with the phylip formatted sequence files either identified as missing or invalid).

I'll use fasta as input for now, or switch to parGenes to generate the trees.

Thanks again,
Andrew

**Comment by BenoitMorel** [comment]
Dear Andrew,

I actually read your messages to fast. The 
``` what(): Could not load open newick file random```
message could have been explain with the previously mentioned bug, but not the issue with the missing/incorrect phylip files.

If you want to share your whole dataset, I'd be happy to try to reproduce the issue on our machines. Even if the problem comes from the dataset, GeneRax/SpeciesRax should be able to give an understandable error message...

I think that running ParGenes before SpeciesRax is usually a better practice, especially because inferring the gene trees often takes more time than inferring the species tree, so it's good to separate the two steps. I will write something about this in the wiki.

**Comment by ak-watson** [comment]
Dear Benoit, 

Happy to share if it helps. Will try to remember to forward when I'm in the lab.

I noticed another potential issue but very much based on user error.

I accidentally tried to run generax with mpiexec from the wrong directory (I'd forgotten i was still inside the mapping dir).

There was no families.txt file in that directory, but 3000 or so mapping files (named X.mapping.tsv).

Rather than quitting cleanly with an error message, the RAM and SWAP are both filled completely before everything crashes out.

Best,
Andrew

**Comment by BenoitMorel** [comment]
Hi Andrew,

thanks, I'll wait for your files.

Indeed, there seems to be something going wrong when specifying the wrong family file path and using more than one MPI rank... Thanks for your report, I'll fix it :-)

Best,
Benoit

**Comment by BenoitMorel** [comment]
I just fixed the invalid family path dirty crash in the master branch (issue #56)

**Comment by ak-watson** [comment]
Sorry for the delayed response, I have been trying a few things to see if I could resolve the issue. 

Briefly, I reformatted my data to name sequences with the >GENE_SPECIES format so that mapping files aren't required, generated trees from fasta formated alignments using ParGenes.

ParGenes seemed to run fine (although i struggled to get the hpc version to work so had to stick to running it on a single node). 

I generated my families.txt file with the latest version of pargenes_to_families_file.py.

I tried to run the latest version of generax using the command in my original message. 

Unfortunately I am still seeing the same error message, that ~500 of ~3200 alignment files are being filtered (the first saying the alignment is invalid, the rest saying the file isn't present).

Then, after the families are discarded I reach  [00:00:21] Starting species tree initialization...

The job is always quickly killed with RAM use maxed out.

To check my system memory wasn't too limited (64Gb) I tried testing this on our HPC with 750Gb allocated. Unfortunately I hit the memory limit again. Perhaps that is just because my dataset is too big? (3284 families, 486205 sequences).

I had one thought. Would the same sequence ID being present in multiple families potentially cause issues?

**Comment by BenoitMorel** [comment]
Hey Andrew,

In addition, here is something you can try, to try to get rid of the sequences issue: do not provide the alignments in the family file. SpeciesRax does not need the alignments to infer the species tree. However, they are needed if you want to do gene tree correction from the species tree. To disable gene tree correction, you need to add `--strategy SKIP`. Could you please try it, even if you are interested in correcting the gene trees? It might help understanding the problem.
You can comment [this line](https://github.com/BenoitMorel/GeneRax/blob/master/scripts/pargenes_to_families_file.py#L54) in the pargenes_to_families_file.py script and run in again.

Also, did ParGenes manage to successfully infer all (or most of) the gene trees?

Regarding memory issues: the memory requirements is roughly a function of (n*m) where n is the number of species and m the total number of gene copies (486205 in your case). Your "silly mistake" could well explain why you hit the memory limit (the n*m could become m*m and then you can indeed expect a crash :-)). 

If you still encounter issues (either hitting the memory limit or having many families discarded for no good reason), just send me your dataset. If I can reproduce, fixing the issue(s) should not be too complicated. 

Best,
Benoit
